Design/histogram_type.md

```diff
@@ -0,0 +1,251 @@
+- Jon pointed out that it may make most sense to default to storing data as `Frequecies`, as OpenGenie did by default.
+  Many operations could probably done in this representation, and furthermore the conversion to `Counts` is cheaper, since it includes a multiplication instead of addition.
+- The current copy-on-write (COW) mechanism seems useful and should be preserved.
+  - The `typedef` `MantidVecPtr` for a `cow_ptr` is useless and confusing and should be removed.
+  - We should consider extending the COW mechanism and make sure sharing is preserved more than it is now. For example, what can we do to maintain sharing of `X` for all histograms in a workspace where `X` is modified in an identical way? The COW mechanism as it is now should break down in that case. Probably this needs to be handled at the workspace/algorithm level, and cannot be dealt with this here? Having parametric `X` could circumvent that in certain cases (see outlook).
+
+#### Interface
+
+- The interface of `Histogram` can be (partially) forwarded by `ISpectrum` and `MatrixWorkspace`, similarly to how it is implemented currently for, e.g., `ISpectrum::dataX` and `MatrixWorkspace::dataX`.
+
+- The main challenge in the design of the `Histogram` interface is the way how we deal with references.
+  If the internal data storage mode of, say, the Y-data is not known, it is not trivial to provide `Counts &counts()` (as a partial equivalent to the current `std::vector<double> &dataY()`), since there may be nothing to reference to.
+  - One way of dealing with this is a conversion of the underlying type on write, e.g.,
+
+    ```cpp
+    // Default storage mode is `Frequencies`.
+    Histogram hist;
+    // Const accessors return by value:
+    // const Counts Histogram::constCounts() const;
+    const auto &counts = hist.constCounts();
+    // Non-const accessors return by reference and converts internal storage mode:
+    // Counts &Histogram::counts();
+    auto &writableCounts = hist.counts();
+    // modify writableCounts
+    ```
+
+  - Alternatively, we could eliminate the option of getting non-const references, and instead do everything via setters, e.g.,
+
+    ```cpp
+    // Default storage mode is `Frequencies`.
+    Histogram hist;
+    // Const accessors return by value (there is no non-const overload):
+    // Counts Histogram::counts() const;
+    auto counts = hist.counts();
+    // modify counts
+    hist.setCounts(counts);
+    ```
+
+  Both options involve some overhead when writing (in the former case we can move the underlying `std::vector<double>` from `Frequencies` to `Counts`, in the latter case we allocate memory for the temporary `Counts` object).
+
+  It is not entirely clear how we would deal with the similar situation in the `BinEdges` vs. `Points` case, where a conversion is not lossless and should thus not be done automatically in the generic case.
```

Comment:
Well, the point data vector is 1 shorter than the bin edge data. When Mantid converts PointData to HistogramData it uses the midpoint between two points. For constant bin sizes that is correct, in all other cases your bin edges or points would slowly drift if you keep converting back and forth.